### PR TITLE
FI-469 Integrate BDT

### PR DESCRIPTION
**Pull request by arscan**
This integrates the Bulk Data Test tool as sequences into Inferno.  We will require the use of Docker in order to use this because it simplifies the configuration substantially.   Docker launches the [Inferno BDT Service](/inferno-community/inferno-bdt-service), which is a service wrapper around the BDT library and executes a number of bulk data tests.

To run:

```
docker-compose build
docker-compose up

# Navigate to http://localhost:4567 ****and force refresh your browser****
```

The force refresh is required because we made a small update to the javascript and haven't yet incremented the version number, which is what usually invalidates our javascript caching.

Choose the `SMART Bulk Server` vendor configuration, which will load the Community tests for this tool.  All tests should pass with the provided configuration (which runs [this configuration](    https://bulk-data.smarthealthit.org/?auth_type=jwks&jwks=eyJrZXlzIjpbeyJrdHkiOiJFQyIsImNydiI6IlAtMzg0IiwieCI6InRWekRXUTVxTTFla1JrTzlhY1YyT3JEV19KeWVTNmJhRHJYUHM5dlBPR094YTV6NTVDV3Rwd2dfQkR5T1l0OG4iLCJ5IjoiaHBYSUNaTU5lekQxb01OYWtPcWxHTmgzV1FnQ2RLZlFVcERObGh4X2RLUW8weW1NUGVYS3hpc01tVnFXT2stTCIsImtleV9vcHMiOlsidmVyaWZ5Il0sImV4dCI6dHJ1ZSwia2lkIjoiYzk1MjlhMGFhZWZhNWVhNjU5ZDY1YzQ5MzllY2E5NzYiLCJhbGciOiJFUzM4NCJ9LHsia3R5IjoiRUMiLCJjcnYiOiJQLTM4NCIsImQiOiJtOWdKcUdhaDhDOC12bW9zNS02YnA5MDJUUEZHSHZjdFVJekdkMjJMUHlFNm81RDJXRVUxbVdFbTc0bEVleWJQIiwieCI6InRWekRXUTVxTTFla1JrTzlhY1YyT3JEV19KeWVTNmJhRHJYUHM5dlBPR094YTV6NTVDV3Rwd2dfQkR5T1l0OG4iLCJ5IjoiaHBYSUNaTU5lekQxb01OYWtPcWxHTmgzV1FnQ2RLZlFVcERObGh4X2RLUW8weW1NUGVYS3hpc01tVnFXT2stTCIsImtleV9vcHMiOlsic2lnbiJdLCJleHQiOnRydWUsImtpZCI6ImM5NTI5YTBhYWVmYTVlYTY1OWQ2NWM0OTM5ZWNhOTc2IiwiYWxnIjoiRVMzODQifV19&page=1000&stu=4)).

Tests are generated by running `bundle exec rake inferno:generate[bdt,bdt]`.  This has stored a local copy of the test index, pulled from the BDT service, which is essentially a JSON file with all the metadata of every test.  If the BDT service is updated, we need grab a copy of the test index and regenerate.

Pull requests into Inferno require the following items to be completed. Submitter and reviewer 
should check the relevant check boxes when the associated item is done. For items that are not 
applicable, note it's not applicable ("N/A") and check the box. For example, external Pull 
Requests do not require ticket links.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Internal ticket for this PR:
- [x] Internal ticket links to this PR
- [x] Internal ticket is properly labeled (Community/Program)
- [x] Internal ticket has a justification for its Community/Program label
- [x] Code diff has been reviewed for extraneous/missing code
- [ ] Tests are included and test edge cases
- [x] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

**Comment by Jammjammjamm** [comment]
The run buttons in the BDT tests don't work for me. I get the following error when I click one:
Screen Shot 2019-11-19 at 8 51 18 AM

**Comment by arscan** [comment]
Sorry!  Please force refresh your browser because I had to update the javascript, and we haven't yet incremented the version number.  I updated the instructions in the description as well.

> The run buttons in the BDT tests don't work for me. I get the following error when I click one:
> Screen Shot 2019-11-19 at 8 51 18 AM

**Comment by Jammjammjamm** [comment]
Using the preset now, all of the tests fail due to a redirect:
Screen Shot 2019-11-20 at 8 37 20 AM

**Comment by hadleynet** [comment]
Not sure I understand the changes to the docker-compose.yml file. Why remove nginx? Does BDT need to be exposed or is it wrapped by inferno and not exposed directly?

**Comment by arscan** [comment]
> Not sure I understand the changes to the docker-compose.yml file. Why remove nginx? Does BDT need to be exposed or is it wrapped by inferno and not exposed directly?

No, BDT does not need to be accessed directly.  In "production" we don't use this docker-compose.yml, we use the docker setup in http://github.com/onc-healthit/inferno-site-overlay.  I removed it here because it was only overhead in a development environment and didn't seem to add anything useful.  But maybe it is good to have it just to be thorough?

**Comment by hadleynet** [comment]
> > Not sure I understand the changes to the docker-compose.yml file. Why remove nginx? Does BDT need to be exposed or is it wrapped by inferno and not exposed directly?
> 
> No, BDT does not need to be accessed directly. In "production" we don't use this docker-compose.yml, we use the docker setup in [http://github.com/onc-healthit/inferno-site-overlay](https://github.com/onc-healthit/inferno-site-overlay). I removed it here because it was only overhead in a development environment and didn't seem to add anything useful. But maybe it is good to have it just to be thorough?

I think the docker-compose.yml should reflect how we think somebody who wants to deploy the code themselves using Docker should do it. Not sure if nginx is overkill for a limited use local install but it wouldn't hurt.

Either way, I don't think you need the bdt_server.ports entry since that is used to expose the bdt_service to the host when it only needs to be used by ruby_server. You should also probably add a ruby_server.depends_on that points to bdt_server.

File: generator/bdt/bdt_generator.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      end
+
+      def copy_base
+        source_file = File.join(__dir__, 'templates/bdt_base.rb')
+        out_file_name = sequence_out_path + '/bdt_base.rb'
+        FileUtils.cp(source_file, out_file_name)
+      end
+
+      # This organizes into a single bulk data group
+      # With a sequence per
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/bdt/bdt_generator.rb]: Fixed

File: config.yml [hunk 2]
```diff
@@ -69,5 +72,22 @@ presets:
...
+    bulk_fastest_resource: Patient
+    bulk_system_export_endpoint: /$export
+    bulk_patient_export_endpoint: /Patient/$export
+    bulk_group_export_endpoint: /Group/5/$export
+    #https://bulk-data.smarthealthit.org/?auth_type=jwks&jwks=eyJrZXlzIjpbeyJrdHkiOiJFQyIsImNydiI6IlAtMzg0IiwieCI6InRWekRXUTVxTTFla1JrTzlhY1YyT3JEV19KeWVTNmJhRHJYUHM5dlBPR094YTV6NTVDV3Rwd2dfQkR5T1l0OG4iLCJ5IjoiaHBYSUNaTU5lekQxb01OYWtPcWxHTmgzV1FnQ2RLZlFVcERObGh4X2RLUW8weW1NUGVYS3hpc01tVnFXT2stTCIsImtleV9vcHMiOlsidmVyaWZ5Il0sImV4dCI6dHJ1ZSwia2lkIjoiYzk1MjlhMGFhZWZhNWVhNjU5ZDY1YzQ5MzllY2E5NzYiLCJhbGciOiJFUzM4NCJ9LHsia3R5IjoiRUMiLCJjcnYiOiJQLTM4NCIsImQiOiJtOWdKcUdhaDhDOC12bW9zNS02YnA5MDJUUEZHSHZjdFVJekdkMjJMUHlFNm81RDJXRVUxbVdFbTc0bEVleWJQIiwieCI6InRWekRXUTVxTTFla1JrTzlhY1YyT3JEV19KeWVTNmJhRHJYUHM5dlBPR094YTV6NTVDV3Rwd2dfQkR5T1l0OG4iLCJ5IjoiaHBYSUNaTU5lekQxb01OYWtPcWxHTmgzV1FnQ2RLZlFVcERObGh4X2RLUW8weW1NUGVYS3hpc01tVnFXT2stTCIsImtleV9vcHMiOlsic2lnbiJdLCJleHQiOnRydWUsImtpZCI6ImM5NTI5YTBhYWVmYTVlYTY1OWQ2NWM0OTM5ZWNhOTc2IiwiYWxnIjoiRVMzODQifV19&page=1000&stu=4
```
Comment (Jammjammjamm) [inline config.yml]: Do we need this comment?

File: lib/app/models/module.rb [hunk 3]
```diff
@@ -4,18 +4,20 @@
...
 require_relative 'module/test_group'
 require_relative 'module/test_set'
+require_relative 'module/tag'
 
 module Inferno
   class Module
     @modules = {}
 
-    attr_accessor :default_test_set
+    attr_accessor :name
```
Comment (Jammjammjamm) [inline lib/app/models/module.rb]: It would be nice to keep these in alphabetical order.

Comment (arscan) [reply lib/app/models/module.rb]: Ah, I didn't realize that was the reasoning behind the order.  I reordered to match the assignments that occurred in initialize, but I'm good with doing alphabetical here.  Fixed.

File: config.yml (hunk 2 above)
Comment (arscan) [reply config.yml]: This link allows you to view the configuration in their UI, which is useful if you want to tweak it at all.  I added a bit more of an explanation.

**Comment by arscan** [comment]
I changed the docker config back to include nginx, and tried to streamline it a bit @hadleynet.  If this is still not optimal we can address separate from this PR.

**Comment by hadleynet** [comment]
> I changed the docker config back to include nginx, and tried to streamline it a bit @hadleynet. If this is still not optimal we can address separate from this PR.

Looks great to me.

**Pull request merged**